From cc5fd9916a630cacb30bed2474bcf5df6c5801ae Mon Sep 17 00:00:00 2001
From: Kevin Eyssartier <kevin.eyssartier@thalesgroup.com>
Date: Tue, 7 Dec 2021 10:31:12 +0100
Subject: [PATCH 02/15] cva6: Add cv32a6 support (#7)

* 32bits configuration
* Adding 32bits configuration

* Bypassing relocation in function boot_relocat_fdt with fdt_high=-1

the hacky workaround is problematic in 32bits, it needs to be investigated

Signed-off-by: Kevin Eyssartier <kevin.eyssartier@thalesgroup.com>

* Device tree and Bootcmd correction
---
 arch/riscv/dts/Makefile                       |   2 +-
 arch/riscv/dts/cv32a6_genesysII.dts           | 145 ++++++++++++++++++
 ...va6_genesysII.dts => cv64a6_genesysII.dts} |   0
 board/openhwgroup/cva6_genesysII/Kconfig      |   3 +-
 .../openhwgroup_cv32a6_genesysII_defconfig    |  37 +++++
 ...=> openhwgroup_cv64a6_genesysII_defconfig} |   4 +-
 6 files changed, 187 insertions(+), 4 deletions(-)
 create mode 100644 arch/riscv/dts/cv32a6_genesysII.dts
 rename arch/riscv/dts/{cva6_genesysII.dts => cv64a6_genesysII.dts} (100%)
 create mode 100644 configs/openhwgroup_cv32a6_genesysII_defconfig
 rename configs/{openhwgroup_cva6_genesysII_defconfig => openhwgroup_cv64a6_genesysII_defconfig} (78%)

diff --git a/arch/riscv/dts/Makefile b/arch/riscv/dts/Makefile
index 2744f5948c..ba246009d4 100644
--- a/arch/riscv/dts/Makefile
+++ b/arch/riscv/dts/Makefile
@@ -6,7 +6,7 @@ dtb-$(CONFIG_TARGET_QEMU_VIRT) += qemu-virt.dtb
 dtb-$(CONFIG_TARGET_SIFIVE_UNLEASHED) += hifive-unleashed-a00.dtb
 dtb-$(CONFIG_TARGET_SIFIVE_UNMATCHED) += hifive-unmatched-a00.dtb
 dtb-$(CONFIG_TARGET_SIPEED_MAIX) += k210-maix-bit.dtb
-dtb-$(CONFIG_TARGET_OPENHWGROUP_CVA6_GENESYSII) += cva6_genesysII.dtb
+dtb-$(CONFIG_TARGET_OPENHWGROUP_CVA6_GENESYSII) += $(addsuffix .dtb, $(CONFIG_DEFAULT_DEVICE_TREE:"%"=%))
 
 targets += $(dtb-y)
 
diff --git a/arch/riscv/dts/cv32a6_genesysII.dts b/arch/riscv/dts/cv32a6_genesysII.dts
new file mode 100644
index 0000000000..41640be51f
--- /dev/null
+++ b/arch/riscv/dts/cv32a6_genesysII.dts
@@ -0,0 +1,145 @@
+/dts-v1/;
+
+/ {
+  #address-cells = <2>;
+  #size-cells = <2>;
+  compatible = "eth,ariane-bare-dev";
+  model = "eth,ariane-bare";
+  chosen {
+    stdout-path = "/soc/uart@10000000:115200";
+    //tick-timer =  "/cpus/cpu@0";
+  };
+  cpus {
+    #address-cells = <1>;
+    #size-cells = <0>;
+    timebase-frequency = <25000000>; // 25 MHz
+    CPU0: cpu@0 {
+      clock-frequency = <50000000>; // 50 MHz
+      device_type = "cpu";
+      reg = <0>;
+      status = "okay";
+      compatible = "eth, ariane", "riscv";
+      riscv,isa = "rv32ima";
+      mmu-type = "riscv,sv32";
+      tlb-split;
+      // HLIC - hart local interrupt controller
+      CPU0_intc: interrupt-controller {
+        #interrupt-cells = <1>;
+        interrupt-controller;
+        compatible = "riscv,cpu-intc";
+      };
+    };
+  };
+  memory@80000000 {
+    device_type = "memory";
+    reg = <0x0 0x80000000 0x0 0x40000000>;
+  };
+  leds {
+    compatible = "gpio-leds";
+    heartbeat-led {
+      gpios = <&xlnx_gpio 1 0>;
+      linux,default-trigger = "heartbeat";
+      retain-state-suspended;
+    };
+  };
+  L26: soc {
+    #address-cells = <2>;
+    #size-cells = <2>;
+    compatible = "eth,ariane-bare-soc", "simple-bus";
+    ranges;
+    clint@2000000 {
+      compatible = "riscv,clint0";
+      interrupts-extended = <&CPU0_intc 3 &CPU0_intc 7>;
+      reg = <0x0 0x2000000 0x0 0xc0000>;
+      reg-names = "control";
+    };
+    PLIC0: interrupt-controller@c000000 {
+      #address-cells = <0>;
+      #interrupt-cells = <1>;
+      compatible = "riscv,plic0";
+      interrupt-controller;
+      interrupts-extended = <&CPU0_intc 11 &CPU0_intc 9>;
+      reg = <0x0 0xc000000 0x0 0x4000000>;
+      riscv,max-priority = <7>;
+      riscv,ndev = <30>;
+    };
+//  debug-controller@0 {
+//    compatible = "riscv,debug-013";
+//    interrupts-extended = <&CPU0_intc 65535>;
+//    reg = <0x0 0x0 0x0 0x1000>;
+//    reg-names = "control";
+//  };
+    uart@10000000 {
+      compatible = "ns16750", "ns16550";
+      reg = <0x0 0x10000000 0x0 0x1000>;
+      clock-frequency = <50000000>;
+      current-speed = <115200>;
+      interrupt-parent = <&PLIC0>;
+      interrupts = <1>;
+      reg-shift = <2>; // regs are spaced on 32 bit boundary
+      reg-io-width = <4>; // only 32-bit access are supported
+    };
+    timer@18000000 {
+      compatible = "pulp,apb_timer";
+      interrupts = <0x00000004 0x00000005 0x00000006 0x00000007>;
+      reg = <0x00000000 0x18000000 0x00000000 0x00001000>;
+      interrupt-parent = <&PLIC0>;
+      reg-names = "control";
+    };
+    xps-spi@20000000 {
+      compatible = "xlnx,xps-spi-2.00.b", "xlnx,xps-spi-2.00.a";
+      #address-cells = <1>;
+      #size-cells = <0>;
+      interrupt-parent = <&PLIC0>;
+      interrupts = < 2 2 >;
+      reg = < 0x0 0x20000000 0x0 0x1000 >;
+      fifo-size = <256>;
+      xlnx,family = "kintex7";
+      xlnx,fifo-exist = <0x1>;
+      xlnx,num-ss-bits = <0x1>;
+      xlnx,num-transfer-bits = <0x8>;
+      xlnx,sck-ratio = <0x4>;
+
+      mmc@0 {
+        compatible = "mmc-spi-slot";
+        reg = <0>;
+        spi-max-frequency = <12500000>;
+        voltage-ranges = <3300 3300>;
+        disable-wp;
+      };
+
+      // mmc-slot@0 {
+      //   compatible = "fsl,mpc8323rdb-mmc-slot", "mmc-spi-slot";
+      //   reg = <0>;  //Chip select 0
+      //   spi-max-frequency = <12500000>;
+      //   voltage-ranges = <3300 3300>;
+      //   //interrupts = < 2 2 >;
+      //   //interrupt-parent = <&PLIC0>;
+      // };
+    };
+    eth: lowrisc-eth@30000000 {
+      compatible = "lowrisc-eth";
+      device_type = "network";
+      interrupt-parent = <&PLIC0>;
+      interrupts = <3 0>;
+      local-mac-address = [00 18 3e 02 e3 7f]; // This needs to change if more than one GenesysII on a VLAN
+      reg = <0x0 0x30000000 0x0 0x8000>;
+    };
+    xlnx_gpio: gpio@40000000 {
+      #gpio-cells = <2>;
+      compatible = "xlnx,xps-gpio-1.00.a";
+      gpio-controller ;
+      reg = <0x0 0x40000000 0x0 0x10000 >;
+      xlnx,all-inputs = <0x0>;
+      xlnx,all-inputs-2 = <0x0>;
+      xlnx,dout-default = <0x0>;
+      xlnx,dout-default-2 = <0x0>;
+      xlnx,gpio-width = <0x8>;
+      xlnx,gpio2-width = <0x8>;
+      xlnx,interrupt-present = <0x0>;
+      xlnx,is-dual = <0x1>;
+      xlnx,tri-default = <0xffffffff>;
+      xlnx,tri-default-2 = <0xffffffff>;
+    };
+  };
+};
diff --git a/arch/riscv/dts/cva6_genesysII.dts b/arch/riscv/dts/cv64a6_genesysII.dts
similarity index 100%
rename from arch/riscv/dts/cva6_genesysII.dts
rename to arch/riscv/dts/cv64a6_genesysII.dts
diff --git a/board/openhwgroup/cva6_genesysII/Kconfig b/board/openhwgroup/cva6_genesysII/Kconfig
index 42127206a4..ea47ed70d2 100644
--- a/board/openhwgroup/cva6_genesysII/Kconfig
+++ b/board/openhwgroup/cva6_genesysII/Kconfig
@@ -15,7 +15,8 @@ config SYS_CONFIG_NAME
 config SYS_TEXT_BASE
 	default 0x80200000 if SPL
 	default 0x80000000 if !RISCV_SMODE
-	default 0x80200000 if RISCV_SMODE
+	default 0x80200000 if RISCV_SMODE && ARCH_RV64I
+	default 0x80400000 if RISCV_SMODE && ARCH_RV32I
 
 config SPL_TEXT_BASE
 	default 0x08000000
diff --git a/configs/openhwgroup_cv32a6_genesysII_defconfig b/configs/openhwgroup_cv32a6_genesysII_defconfig
new file mode 100644
index 0000000000..93fe352438
--- /dev/null
+++ b/configs/openhwgroup_cv32a6_genesysII_defconfig
@@ -0,0 +1,37 @@
+CONFIG_RISCV=y
+CONFIG_DEFAULT_DEVICE_TREE="cv32a6_genesysII"
+CONFIG_TARGET_OPENHWGROUP_CVA6_GENESYSII=y
+CONFIG_RISCV_SMODE=y
+# CONFIG_RISCV_ISA_C is not set
+CONFIG_OF_BOARD_SETUP=y
+CONFIG_BOOTDELAY=5
+CONFIG_USE_BOOTCOMMAND=y
+CONFIG_BOOTCOMMAND="mmc info; mmc read 90000000 100000 5000; setenv fdt_high 0xffffffff; bootm 90000000 - $(fdtcontroladdr)"
+CONFIG_DISPLAY_CPUINFO=y
+CONFIG_SPL_SPI_FLASH_MTD=y
+CONFIG_CMD_GPT=y
+# CONFIG_RANDOM_UUID is not set
+CONFIG_CMD_MMC=y
+CONFIG_CMD_MTD=y
+CONFIG_CMD_PART=y
+CONFIG_OF_EMBED=y
+CONFIG_MMC=y
+# CONFIG_MMC_WRITE is not set
+CONFIG_MMC_SPI=y
+CONFIG_DM_MTD=y
+CONFIG_MTD_SPI_NAND=y
+CONFIG_SPI_FLASH_ATMEL=y
+CONFIG_SPI_FLASH_EON=y
+CONFIG_SPI_FLASH_GIGADEVICE=y
+CONFIG_SPI_FLASH_ISSI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_SPI_FLASH_SPANSION=y
+CONFIG_SPI_FLASH_STMICRO=y
+CONFIG_SPI_FLASH_SST=y
+CONFIG_SPI_FLASH_WINBOND=y
+CONFIG_SPI_FLASH_XMC=y
+# CONFIG_SPI_FLASH_USE_4K_SECTORS is not set
+CONFIG_SPI_FLASH_MTD=y
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_XILINX_SPI=y
diff --git a/configs/openhwgroup_cva6_genesysII_defconfig b/configs/openhwgroup_cv64a6_genesysII_defconfig
similarity index 78%
rename from configs/openhwgroup_cva6_genesysII_defconfig
rename to configs/openhwgroup_cv64a6_genesysII_defconfig
index 82c1f32296..9c4f39709f 100644
--- a/configs/openhwgroup_cva6_genesysII_defconfig
+++ b/configs/openhwgroup_cv64a6_genesysII_defconfig
@@ -1,12 +1,12 @@
 CONFIG_RISCV=y
-CONFIG_DEFAULT_DEVICE_TREE="cva6_genesysII"
+CONFIG_DEFAULT_DEVICE_TREE="cv64a6_genesysII"
 CONFIG_TARGET_OPENHWGROUP_CVA6_GENESYSII=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_BOOTDELAY=5
 CONFIG_USE_BOOTCOMMAND=y
-CONFIG_BOOTCOMMAND="mmc info; fdt addr $(fdtcontroladdr); fdt set /chosen stdout-path \"/soc/serial@10000000:115200\"; mmc read 90000000 2000 5000; bootm 90000000 - $(fdtcontroladdr)"
+CONFIG_BOOTCOMMAND="mmc info; mmc read 90000000 100000 5000; setenv fdt_high 0xffffffffffffffff; bootm 90000000 - $(fdtcontroladdr)"
 CONFIG_DISPLAY_CPUINFO=y
 CONFIG_SPL_SPI_FLASH_MTD=y
 CONFIG_CMD_GPT=y
-- 
2.31.1

